internal/vendors/service.go

```diff
@@ -61,6 +61,37 @@ func (v *VendorService) BlastEmail(ctx context.Context, vendorIDs []string, temp
 
 	v.applyDefaultEmailTemplate(&template)
 
+	return v.executeBlastEmail(ctx, vendors, template)
+}
+
+func (v *VendorService) AutomatedEmailBlast(ctx context.Context, productName string) ([]string, error) {
+	vendors, err := v.vendorDBAccessor.BulkGetByProductName(ctx, productName)
+
+	if err != nil {
+		return nil, err
+	}
+
+	template := &emailTemplate{}
+	v.applyDefaultEmailTemplate(template)
+	replacements := map[string]string{
+		"{{product_name}}": productName,
+	}
+
+	template.Body = v.replacePlaceholder(template.Body, replacements)
+
+	return v.executeBlastEmail(ctx, vendors, *template)
+}
+
+func (*VendorService) applyDefaultEmailTemplate(template *emailTemplate) {
+	if template.Subject == "" {
+		template.Subject = "Request for products"
+	}
+	if template.Body == "" {
+		template.Body = "Kepada Yth {{name}},\n\nKami mengajukan permintaan untuk pengadaan produk {{product_name}} yang dibutuhkan oleh perusahaan kami. Mohon informasi mengenai ketersediaan, harga, dan waktu pengiriman untuk produk tersebut.\n\nTerima kasih atas perhatian dan kerjasamanya.\n\nHormat kami"
+	}
+}
+
+func (v *VendorService) executeBlastEmail(ctx context.Context, vendors []Vendor, template emailTemplate) ([]string, error) {
```

Comment:
Why do we create this new function?

Reply:
I think so it can be use by both AutomatedEmailBlast and BlastEmail since the implementation is the same to avoid duplication. CMIIW @valenciusap17

Reply:
agreed

Reply:
Do you think this relates to SOLID programming?
cc: @edutjie @ferryhieronimus

Reply:
I just realized that previously, the `BlastEmail` function was doing ‘too much,’ including fetching vendors by ID, applying templates, and concurrently sending emails. This might violate the Single Responsibility Principle, which states that a class (or in this case, a function) should have only one reason to change. However, this function had multiple reasons to change. To address this, we broke down its functionalities into separate functions, which also improved reusability (DRY) and made debugging easier.

Reply:
Yappp, that is correct guys. The seperation to a reusable function was designed to reduce duplication, creating a seperation of concern, and ultimately increasing the readability of the code